backend/satellite_tools/repo_plugins/yum_src.py

```diff
@@ -462,26 +463,97 @@ def __init__(self, url, name, insecure=False, interactive=True,
+                if url in ['', None]:
+                    continue
+                try:
+                    # This started throwing ValueErrors, BZ 666826
+                    (s,b,p,q,f,o) = urlparse(url)
+                    if p[-1] != '/':
+                        p = p + '/'
+                except (ValueError, IndexError, KeyError) as e:
+                    s = 'blah'
+
+                if s not in ['http', 'ftp', 'file', 'https']:
+                    skipped = url
+                    continue
+                else:
+                    goodurls.append(urlunparse((s,b,p,q,f,o)))
+            return goodurls
+
+        try:
+           with open(mirrorlist_path, 'r') as mirrorlist_file:
+               content = mirrorlist_file.readlines()
+        except Exception as exc:
+            self.error_msg("Could not read mirrorlist: {}".format(exc))
+
+        try:
+            # Try to read a metalink XML
+            for files in etree.parse(mirrorlist_path).getroot():
+                file_elem = files.find(METALINK_XML+'file')
+                if file_elem.get('name') == 'repomd.xml':
+                    _urls = file_elem.find(METALINK_XML+'resources').findall(METALINK_XML+'url')
+                    for _url in _urls:
+                        # The mirror urls in the metalink file are for repomd.xml so it
+                        # gives a list of mirrors for that one file, but we want the list
+                        # of mirror baseurls. Joy of reusing other people's stds. :)
+                        if not _url.text.endswith("/repodata/repomd.xml"):
+                            continue
+                        returnlist.append(_url.text[:-len("/repodata/repomd.xml")])
+        except Exception as exc:
+            # If no metalink XML, we try to read a mirrorlist
+            for line in content:
+                if re.match('^\s*\#.*', line) or re.match('^\s*$', line):
+                    continue
+                mirror = re.sub('\n$', '', line) # no more trailing \n's
+                (mirror, count) = re.subn('\$ARCH', '$BASEARCH', mirror)
+                returnlist.append(mirror)
+
+        returnlist = _replace_and_check_url(returnlist)
+
+        try:
+           # Write the final mirrorlist that is going to be pass to Zypper
+           with open(mirrorlist_path, 'w') as mirrorlist_file:
+               mirrorlist_file.write(os.linesep.join(returnlist))
+        except Exception as exc:
+            self.error_msg("Could not write the calculated mirrorlist: {}".format(exc))
+        return returnlist
+
     def setup_repo(self, repo):
         """
         Setup repository and fetch metadata
         """
         self.zypposync = ZyppoSync(root=repo.root)
         zypp_repo_url = self._prep_zypp_repo_url(self.url)
 
+        mirrorlist = self._get_mirror_list(repo, zypp_repo_url)
+        repo.baseurl = repo.baseurl + mirrorlist
```

Comment:
Is this correct? AFAIK `mirrorlist` is a list of urls. Is it correct to concat them with a string?
Or is this a special handling from python?

Reply:
Yep, in this case `repo.baseurl` is actually a `list`: https://github.com/uyuni-project/uyuni/blob/bc2a1fd70fe657e3703be7213c1df92d1d357606/backend/satellite_tools/repo_plugins/yum_src.py#L146-L149